test/studies/ml/lib/Tensor.chpl

```diff
@@ -0,0 +1,925 @@
+            this.rank = other.rank;
+            this.eltType = other.eltType;
+            this._domain = other._domain;
+            this.data = other.data;
+        }
+
+        operator =(ref lhs: Tensor(?rank,?eltType), rhs: Tensor(rank,eltType)) {
+            lhs._domain = rhs._domain;
+            lhs.data = rhs.data;
+        }
+
+        operator =(ref lhs: Tensor(?rank,?eltType), rhs: [?d] eltType) where d.rank == rank {
+            lhs._domain = d;
+            lhs.data = rhs;
+        }
+
+        proc init=(rhs: [?d] eltType) where d.rank == rank {
+            this.init(d.rank,eltType);
+            this.reshapeDomain(d);
+            this.data = rhs;
+        }
+
+        operator :(from: [?d] ?eltType, type toType: Tensor(d.rank,eltType)) {
+            var t: Tensor(d.rank,eltType) = from;
+            return t;
+        }
+        
+        // ------------ Wasnt sure what to do with these   ----------------
+
+        /*operator =(ref lhs: Tensor(?rank,?eltType), in rhs: ?it) where (isRefIterType(it) || (isArray(rhs) && rhs.eltType == eltType)) && rhs.rank == rank {
+            lhs.reshapeDomain(rhs.domain);
+            lhs.data = rhs;
+        }
+        proc init=(in rhs: ?it) where isRefIterType(it) || isArray(rhs) {
+            this.init(rank,eltType);
+            this.reshapeDomain(rhs.domain);
+            this.data = rhs;
+        }
+        operator :(in from: ?it, type toType: Tensor(?rank,?eltType)) where (isRefIterType(it) || (isArray(from) && from.eltType == eltType)) && from.rank == rank {
+            // compilerError("Cannot convert from ",from.type:string," to ",toType:string);
+            var t: Tensor(rank,eltType) = from;
+            return t;
+        }*/
+
+        operator =(ref lhs: Tensor(?rank,?eltType), itr) where itr.type:string == "promoted expression" || itr.type:string == "iterator" {
+            lhs.reshapeDomain(itr.domain);
+            lhs.data = itr;
+        }
+
+        proc init=(itr) where itr.type:string == "promoted expression" || itr.type:string == "iterator" {
+            const A = itr;
+            this.init(A);
+        }
+
+        operator :(itr, type toType: Tensor(?rank,?eltType)) where itr.type:string == "promoted expression" || itr.type:string == "iterator" {
+            var t: Tensor(rank,eltType) = itr;
+            return t;
+        }
+
+
+        // Transposes a vector to a row matrix
+        proc transpose() where rank == 1 {
+            const (p,) = shape;
+            var t = new Tensor(eltType,1,p);
+            t.data[0,..] = this.data;
+            return t;
+        }
+
+        // Matrix transpose
+        proc transpose() where rank == 2 {
+            const (m,n) = this.shape;
+            var M = new Tensor(2,eltType);
+            M.reshapeDomain({0..#n,0..#m});
+            forall (i,j) in M.domain with (ref M, ref this) {
+                M.data[i,j] = this.data[j,i];
+            }
+            return M;
```

Comment:
Although convenient, I'm worried that these transpose operations are pretty inefficient as they involve allocating a new tensor and doing a bunch of copies (problematic when called in a loop). I wonder if it would be better to have a procedure called `transposeMultiply` (or something similar) that does a transpose and multiplication in one operation. I.e., change expressions like:
```chapel
const newDelta = weights.transpose() * delta;
```
into:
```chapel
const newDelta = weights.transposeMultiply(delta);
```

In the implementation, you'd simply index into the matrix as if it were transposed (i.e., switch `i` and `j`), and avoid allocating a temporary Tensor.

Reply:
Yes, I can do this. I was focusing on readability of the code